apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClient.cs

```diff
@@ -0,0 +1,303 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Compute.v1.Data;
+using Google.Apis.Http;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Metadata.V1
+{
+    /// <summary>
+    /// Abstract class providing operations for using metadata for Google Compute Engine.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// This abstract class is provided to enable testability while permitting
+    /// additional operations to be added in the future. You can construct a <see cref="MetadataClientImpl"/> directly.
+    /// </para>
+    /// <para>
+    /// All instance methods declared in this class are virtual, with an implementation which simply
+    /// throws <see cref="NotImplementedException"/>. All these methods are overridden in <see cref="MetadataClientImpl"/>.
+    /// </para>
+    /// </remarks>
+    public abstract class MetadataClient
```

Comment:
How do you plan on instantiating the `Impl` class? Are you going to have a `Create()` method? This pattern is still not very common on C# so we have to be careful here. The other API libraries do just us `new` to create the services, passing an initializer object. The higher level libraries do use this technique of having multiple 'Impl' classes but they are mostly hidden away from the user as an implementation detail.